multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,325 @@
+			// ResourceImport with stale label identity to ID mapping must be deleted.
+			klog.ErrorS(err, "Failed to delete outdated LabelIdentity kind of ResourceImport for label", "label", label)
+		}
+	}
+	id, err := r.allocator.allocate()
+	if err != nil {
+		klog.ErrorS(err, "Failed to allocate ID for new labels", "label", labelHash)
+		return false
+	}
+	labelIdentityResImport = getLabelIdentityResImport(labelHash, label, namespace, id)
+	if err := r.Client.Create(ctx, labelIdentityResImport, &client.CreateOptions{}); err != nil {
+		klog.ErrorS(err, "Failed to create LabelIdentity kind of ResourceImport for new label", "label", label)
+		r.allocator.release(id)
+		return false
+	}
+	r.labelsToID[labelHash] = id
+	if clusters, ok := r.labelsToClusters[labelHash]; ok {
+		clusters.Insert(clusterID)
+	} else {
+		r.labelsToClusters[labelHash] = sets.NewString(clusterID)
+	}
+	if labels, ok := r.clusterToLabels[clusterID]; ok {
+		labels.Insert(labelHash)
+	} else {
+		r.clusterToLabels[clusterID] = sets.NewString(labelHash)
+	}
+	return true
+}
+
+func getLabelIdentityResImport(labelHash, label, ns string, id uint32) *mcsv1alpha1.ResourceImport {
+	return &mcsv1alpha1.ResourceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      labelHash,
+			Namespace: ns,
+		},
+		Spec: mcsv1alpha1.ResourceImportSpec{
+			Kind: common.LabelIdentityKind,
+			LabelIdentity: &mcsv1alpha1.LabelIdentitySpec{
+				Label: label,
+				ID:    id,
+			},
+		},
+	}
+}
+
+// parseLabelIdentityExportNamespacedName gets the clusterID and label identity
+// hash from the API request.
+func parseLabelIdentityExportNamespacedName(namespacedName types.NamespacedName) (string, string) {
+	lastIdx := strings.LastIndex(namespacedName.Name, "-")
+	clusterID := namespacedName.Name[:lastIdx]
+	labelHash := namespacedName.Name[lastIdx+1:]
+	return clusterID, labelHash
+}
+
+// idAllocator allocates an unqiue uint32 ID for each label identity.
+type idAllocator struct {
+	sync.Mutex
+	maxID                  uint32
+	nextID                 uint32
+	previouslyAllocatedIDs intsets.Sparse
+	releasedIDs            intsets.Sparse
+}
+
+// allocate will first try to allocate an ID within the pool of IDs that has been
+// released (due to label identity deletions). If there's no such IDs, it will
+// then allocate the first ID that has not been pre-allocated, or return an error
+// if all IDs have been exhausted.
+func (a *idAllocator) allocate() (uint32, error) {
+	a.Lock()
+	defer a.Unlock()
+
+	available := -1
+	if ok := a.releasedIDs.TakeMin(&available); ok {
+		return uint32(available), nil
+	}
+	for a.previouslyAllocatedIDs.Has(int(a.nextID)) {
+		a.nextID += 1
+	}
+	if a.nextID <= a.maxID {
+		allocated := a.nextID
+		a.nextID += 1
+		return allocated, nil
+	}
+	return 0, fmt.Errorf("no ID available")
+}
+
+// setAllocated reserves IDs allocated during the previous round of label identity
```

Comment:
I feel it is to reserve IDs allocated before controller restarted.

Reply:
Rephrased